File: Fourth Semester - Design and Programming Environments/Java/ProiectJavaNetworking/TravelModel/src/java/org/example/model/Reservation.java
package org.example.model;

import java.util.Objects;

public class Reservation extends Entity<Integer> {
    private Tour tour;
    private String clientName;
    private String clientContact;
    private Integer numberOfReservedSeats;

    public Reservation(Tour tour, String clientName, String clientContact, Integer numberOfReservedSeats) {
        this.tour = tour;
        this.clientName = clientName;
        this.clientContact = clientContact;
        this.numberOfReservedSeats = numberOfReservedSeats;
    }

    public Tour getTour() {
        return tour;
    }

    public void setTour(Tour tour) {
        this.tour = tour;
    }

    public String getClientName() {
        return clientName;
    }

    public void setClientName(String clientName) {
        this.clientName = clientName;
    }

    public String getClientContact() {
        return clientContact;
    }

    public void setClientContact(String clientContact) {
        this.clientContact = clientContact;
    }

    public Integer getNumberOfReservedSeats() {
        return numberOfReservedSeats;
    }

    public void setNumberOfReservedSeats(Integer numberOfReservedSeats) { this.numberOfReservedSeats = numberOfReservedSeats; }


    @Override
    public String toString() {
        return "Reservation{" +
                "tour=" + tour +
                ", nameClient='" + clientName +
                ", contactClient='" + clientContact +
                ", numberOfReservedPlaces=" + numberOfReservedSeats +
                '}';
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if(!(obj instanceof Reservation reservation)) {
            return false;
        }
        return getTour().equals(reservation.getTour()) && getClientName().equals(reservation.getClientName()) &&
                getClientContact().equals(reservation.getClientContact()) &&
                getNumberOfReservedSeats().equals(reservation.getNumberOfReservedSeats());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getTour(), getClientName(), getClientContact(), getNumberOfReservedSeats());
    }
}

File: Fourth Semester - Design and Programming Environments/C#/ProiectC#Networking/TravelModel/model/Reservation.cs
namespace TravelModel.model;

public class Reservation : Entity<int>
{
    private Tour _tour;
    private String _clientName;
    private String _clientContact;
    private int _numberOfReservedSeats;

    public Reservation(Tour tour, String clientName, String clientContact, int numberOfReservedSeats)
    {
        _tour = tour;
        _clientName = clientName;
        _clientContact = clientContact;
        _numberOfReservedSeats = numberOfReservedSeats;
    }

    public Tour Tour => _tour;

    public String ClientName
    {
        get => _clientName;
        set => _clientName = value;
    }

    public String ClientContact
    {
        get => _clientContact;
        set => _clientContact = value;
    }

    public int NumberOfReservedSeats
    {
        get => _numberOfReservedSeats;
        set => _numberOfReservedSeats = value;
    }

    public override string ToString()
    {
        return _tour.ToString() + $"Client Name: {_clientName}, Client Contact: {_clientContact}, Number of Reserved Seats: {_numberOfReservedSeats}";
    }

    public override bool Equals(object obj)
    {
        if (obj == null || GetType() != obj.GetType())
            return false;
        Reservation other = (Reservation)obj;
        return _tour == other._tour && _clientContact == other._clientContact && _clientName == other._clientName && 
               _numberOfReservedSeats == other._numberOfReservedSeats;

    }

    public override int GetHashCode()
    {
        return HashCode.Combine(_tour, _clientName, _clientContact, _numberOfReservedSeats);
    }
}